### PR TITLE
Add a Swift parser library that is written in Swift

### PR DESCRIPTION
**Pull request by ahoppen**
This starts an effort to implement the parser for the Swift compiler in Swift itself.

For more information, please see the announcement forum post at: https://forums.swift.org/t/a-new-swift-parser-for-swiftsyntax/59813

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Review by CodaFi — APPROVED** [review]

**Review by rintaro — APPROVED** [review]

**Comment by etcwilde** [comment]
@compnerd, if you have a cycle or two, it's probably worth making sure this works on Windows.

**Review by DougGregor — APPROVED** [review]

**Review by zoecarver — APPROVED** [review]

**Comment by compnerd** [comment]
Doesn't seem to build on Windows (which is a regression).  At the very least we will need https://github.com/apple/swift-docc-plugin/pull/26.

**Comment by CodaFi** [comment]
The docc plugin is not required for this repository in general, and is not a dependency for this PR specifically. Docc support is built into Xcode, I am not sure why we have a plugin dependency as well.

**Comment by compnerd** [comment]
I couldn't answer why we have that.  We should invert the condition so that it is not used by default.

**Comment by DougGregor** [comment]
I'm fine with removing the DocC plugin dependency or inverting the condition. We shouldn't land this in a form that breaks the Windows build.

**Comment by CodaFi** [comment]
Let's strike it for now. We shouldn't have merged it without verifying that it worked for Windows

https://github.com/apple/swift-syntax/pull/617

File: Sources/swift-parser-test/swift-parser-test.swift [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+
+import SwiftSyntax
+import SwiftParser
+import Foundation
+import ArgumentParser
```
Comment (compnerd) [inline Sources/swift-parser-test/swift-parser-test.swift:17]:
```suggestion
import ArgumentParser
#if os(Windows)
import WinSDK
#endif
```

File: Sources/swift-parser-test/swift-parser-test.swift [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+      let sema = DispatchSemaphore(value: 0)
+      process.standardOutput = FileHandle.nullDevice
+      process.standardError = FileHandle.nullDevice
+      process.terminationHandler = { process in
+        sema.signal()
+      }
+
+      try process.run()
+      if sema.wait(timeout: DispatchTime.now() + .seconds(1)) == .timedOut {
+        kill(pid_t(process.processIdentifier), SIGKILL)
```
Comment (compnerd) [inline Sources/swift-parser-test/swift-parser-test.swift:232]:
```suggestion
#if os(Windows)
        _ = TerminateProcess(process.processHandle, 0)
#else
        kill(pid_t(process.processIdentifier), SIGKILL)
#endif
```

**Review by compnerd — COMMENTED** [review]
This at least allows this PR to be built on Windows with:
```
swift build -Xlinker -Xcc -IC:\Library\Developer\Toolchains\unknown-Asserts-development.xctoolchain\usr\include
```

Note that building this requires a locally built toolchain as there was a fix for SPM that is needed that has not had a new toolchain snapshot built due to other failures introduced by the rebranch.

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test

File: Sources/SwiftParser/README.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
+## Overview
```
Comment (benlangmuir) [inline Sources/SwiftParser/README.md:1]: (not a blocker) The existence of this file triggers a warning, at least with 5.7:

```
warning: 'swift-syntax': found 1 file(s) which are unhandled; explicitly declare them as resources or exclude from the target
    Sources/SwiftParser/README.md
```

**Comment by wu-s-john** [comment]
Nice. Swift would really benefit from code generating it's own code rather than C++.

File: Sources/SwiftParser/README.md (hunk 3 above)
Comment (DougGregor) [reply Sources/SwiftParser/README.md:1]: I'll merge and we'll fix it up afterward. Thanks!

**Comment by compnerd** [comment]
> I can assure you that this does currently not work on Windows. I just tried using this framework in an SPM project on Windows.

Please use a build from main.  At the very least, a recent snapshot.

**Comment by broken-bytes** [comment]
I used the most recent version.

```dependencies: [
        .package(
            url: "https://github.com/apple/swift-syntax.git", from: "0.50600.1"),
    ],
    ```
    As far as I am concerned, this is the most recent snapshot

**Comment by compnerd** [comment]
5.6 should be buildable - https://github.com/compnerd/swift-build/releases/tag/swift-format-5.6-RELEASE was built and would pull in swift-syntax 5.6.  I wonder if the .1 has anything to do with it 🤔

**Comment by DougGregor** [comment]
@broken-bytes If you're looking to try the new parser that's the subject of this pull request, you'll need `main`.

**Pull request merged**